Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,619 @@
+
+using System;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Data;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.Caching;
+using System.Security;
+using System.Text.RegularExpressions;
+using System.Threading;
+using GSF.Collections;
+using GSF.Configuration;
+using GSF.Data;
+using GSF.Diagnostics;
+using GSF.Identity;
+using JWT;
+using JWT.Algorithms;
+using JWT.Exceptions;
+using JWT.Serializers;
+using Newtonsoft.Json;
+
+#pragma warning disable S2068
+
+namespace GSF.Security
+{
+    /// <summary>
+    /// Represents an <see cref="ISecurityProvider"/> that uses openID Connect
+    /// </summary>
+    /// <example>
+    /// Required config file entries (automatically added):
+    /// <code>
+    /// <![CDATA[
+    /// <?xml version="1.0"?>
+    /// <configuration>
+    ///   <configSections>
+    ///     <section name="categorizedSettings" type="GSF.Configuration.CategorizedSettingsSection, GSF.Core" />
+    ///   </configSections>
+    ///   <categorizedSettings>
+    ///     <securityProvider>
+    ///       <add name="ProviderType" value="GSF.Security.OIDCSecurityProvider, GSF.Security" description="The type to be used for enforcing security."
+    ///         encrypted="false" />
+    ///       <add name="ClientID" value="xxxx-xxxx-xxxx" description="Defines the ClientID as required per OpenID Connect Standard." encrypted="false" />
+    ///       <add name="Scope" value="user" description="Defines the Scope as required per OpenID Connect Standard." encrypted="false" />
+    ///       <add name="AuthorizationEndpoint" value="user" description="Defines the Endpoint to redirect the user for Authorization." encrypted="false" />
+    ///       <add name="RedirectURI" value="https://localhost:8986/" description="Defines the URI the User get's redirected to after signing in." encrypted="false" />
+    ///       <add name="ClientSecret" value="sssss-ssssss-sssss" description="Defines the Client Secret to encrypt User Information." encrypted="false" />
+    ///       <add name="TokenEndpoint" value="user" description="Defines the Endpoint to get the User Token from." encrypted="false" />
+    ///     </securityProvider>
+    ///   </categorizedSettings>
+    /// </configuration>
+    /// ]]>
+    /// </code>
+    /// </example>
+
+    public class OIDCSecurityProvider : SecurityProviderBase
+    {
+        #region [ Members ]
+
+        // Constants
+        private const string ResponseType = "code";                          // Response Type the SecurityProvider expects
+        //private const string ResponseMode = "form_post";                     // The mode used to get ID Token for the end-user
+        private int NonceSlidingExpiration = 600;                            // Time in Seconds that a Nonce is valid for. This prevents replay attacks
+        // Fields
+        private string m_ClientID;
```

Comment:
Use pascal casing for these member variables starting with first letter beyond `m_` being lower case, e.g., `m_clientID`